// TaiwanReservoirAPI/libs/main.js
/* Merge statistic data of resevoir.js with real-time data of reservoir_today.js*/
var reservoir = require('./reservoir'),
    reservoirNow = require('./reservoir_today'),
    reservoirPast = require('./reservoir_past');
var jsonQuery = require('json-query');

module.exports = {

  statistic: function(callback){
    reservoir(function(err,data){
      if(err) callback(err,null);
      callback(null,data);
    });
  },

  immediate: function(callback){
    reservoirNow(function(err,data){
      if(err) callback(err,null);
      callback(null,data);
    });
  },

  getPastStatistic: function(callback, pastDate)
  {
    reservoirPast(function(err,data){
      if(err) callback(err,null);
      callback(null,data);
    },pastDate);
  },

  merge: function(callback){
    reservoir(function (err, reservoirData) {
        if (err)
            callback(err, null);

        reservoirNow(function (err, immediateData) {
            if (err)
                callback(err, null);

            immediateData.forEach(function (element, index, array) {
                var key = jsonQuery('reserv[reservoirName=' + element.reservoirName + ']', {
                    data: {
                        reserv: reservoirData
                    }
                }).key;

                if (key !== null) {
                    reservoirData[key].immediateTime = element.immediateTime;
                    reservoirData[key].immediateLevel = element.immediateLevel;
                    reservoirData[key].immediateStorage = element.immediateStorage;
                    reservoirData[key].immediatePercentage = element.immediatePercentage;
                }

            });

            callback(null, reservoirData);
        });
    });
  }
};
